### ISSUE TITLE
Open pane scrollback in default editor

### ISSUE THREAD

**Issue by imsnif** [zellij-org/zellij#1403]
The experience should be:
1. Press a Zellij shortcut (something like ctrl+s + e)
2. Have the current pane replaced with your default editor, opened to a temporary file containing all the scrollback of the terminal pane and scrolled to the same line
3. Search, save, copy/paste, do whatever you want
4. When you quit the editor (or press ctrl+s + e again), the original pane comes back

Depends on: https://github.com/zellij-org/zellij/pull/1375

**Comment by imsnif** [comment]
@cosminadrianpopescu - you mentioned you might like to work on this. I wanted to open this as a separate issue because it's part of our roadmap.

**Comment by cosminadrianpopescu** [comment]
Yes, I would like to work on this.

**Comment by cosminadrianpopescu** [comment]
How would I go about it? What is the idea? How do I replace the content of the pane with a command without putting that command in the history of the current pane's shell?

Also, when running the command, is there something to know about it regarding different operating systems?

My idea would be to create a new temporary pane which would replace the current one. Is there a better approach? If this is the right approach, how do I know when this temporary panel is closed, to react and put back the old pane?

Any hints?

**Comment by imsnif** [comment]
I think the basic approach should be:

1. When we receive the `EditScrollback` action (which we should create), we first dump the scrollback of the current terminal pane into a temporary file. This is something we do in the `screen_thread`, in `Tab` in a similar function to the one you already know here: https://github.com/zellij-org/zellij/blob/main/zellij-server/src/tab/mod.rs#L1385
2. Then, we need to open a pane with the default editor and point it to that file. In our `pty_thread` we already have a facility for that. So if we send it the right instruction (`PtyInstruction::SpawnTerminal` with the `TerminalAction::OpenFile(path)` parameter), it'll do all this work for us. Here's an example where our plugins send it this instruction: https://github.com/zellij-org/zellij/blob/c53b3235c848bbf1c6f56e4c00f46e99b47f9cd6/zellij-server/src/wasm_vm.rs#L395-L399
3. The above function will send back a message to `Screen` (and eventually to the relevant `Tab`) once the pane is open, giving us the pane ID. The problem is that it'll send the message to us as a new pane and we'll open it as a new pane instead of replacing it with the current pane. 
4. What I suggest we do is create a new sort of `PtyInstruction` instead of `PtyInstruction::SpawnTerminal` (https://github.com/zellij-org/zellij/blob/main/zellij-server/src/pty.rs#L85). Maybe something like `PtyInstruction::OpenInplaceEditor` which will act in the same way as `PtyInstruction::SpawnTerminal`, but carry with it a bit of extra context that we can use in order to identify the pane we're operating on. We could use the temporary file name to where we wrote the scrollback in order to identify it (I'm counting on these being unique throughout the system).
5. So we send the new `OpenInplaceEditor` instruction to the pty_thread, and we keep a temporary filename on our `Tab` state... maybe in a `HashMap<file_name, original_pane_id>` for later. The pty_thread sends us a new kind of `ScreenInstruction` back (instead of `ScreenInstruction::NewPane` - https://github.com/zellij-org/zellij/blob/main/zellij-server/src/pty.rs#L89) - maybe something like `ScreenInstruction::OpenInplaceEditor` which will contain everything `NewPane` contains, but also the temporary file so we can identify it in `Tab`.
6. Once that's done, in `Screen` we handle this new instruction, thread it down to the relevant `Tab` which will then:
  - replace the pane's `geom` with the `geom` of the pane we're replacing
  - place the pane we're replacing in `self.panes_to_hide` (this is the strategy we use for placing a pane in full screen - so we can safely reuse it here)
  - render everything and we're done
7. For closing the pane, we'll need to keep another bit of state on `Tab`, maybe  a `HashMap<scrollback_pane_id, originating_pane_id>`, so that we know to take the pane our of `self.panes_to_hide` instead of going through our whole normal close_pane jam.

Makes sense? Shall I go into more details about specific sections? A lot of the questions you asked are already handled by the app itself as you can see above. It's just a question of knowing where to look. And since this is a bit of an involved (and in need of refactoring) app, I'm happy to guide you as closely as you'd like me to :)

**Comment by imsnif** [comment]
Hey @cosminadrianpopescu - sorry to press on this, I just want to release this feature with the next release in ~1 week. Do you think you can see it to completion, or are you too busy?

**Comment by cosminadrianpopescu** [comment]
I am planning to work on this from Wednesday. I hope to finish it until Sunday. I will keep you updated. 

I looked already at it. It seems straight forward what you described. What I'm most worried is sending those information back and forth when I want to see what panel is opened (points 5 and 7 in your tutorial). I remember the difficulties I've had in the test case of the editor dump feature. Other than this, it should be straight forward and I should be able to finish it until Sunday.

**Comment by imsnif** [comment]
Sounds great, thanks for the quick response!

If you have trouble, I'm also happy to work on this together. You can also give me a ping on Discord if you like.

**Comment by cosminadrianpopescu** [comment]
I am at step 6 and having some issues. So far, what I've come up with is this:

```rust
            ScreenInstruction::OpenInPlaceEditor(pid, file_name, client_id) => {
                if let Some(active_tab) = screen.get_active_tab_mut(client_id) {
                    if let Some(pane1) = active_tab.get_active_pane(client_id) {
                        scrollbacks.insert(file_name, pane1.pid());
                        let geom = pane1.current_geom().clone();
                        active_tab.new_pane(pid, Some(client_id));
                        if let Some(pane2) = active_tab.get_active_pane_mut(client_id) {
                            pane2.set_geom(geom);
                        }
                        else {
                            log::error!("New editor pane not found. Maybe there is an issue with launching it?: {:?}", client_id);
                        }
                    }
                } else {
                    log::error!("Active tab not found for client id: {:?}", client_id);
                }
                screen
                    .bus
                    .senders
                    .send_to_server(ServerInstruction::UnblockInputThread)
                    .unwrap();
                screen.update_tabs();

                screen.render();
            }
```

I have 2 issues: 

1. The panel is hidden and replaced with the editor buffer, but the geometry is not right. The editor panel has only half of the screen, like it would have if the original buffer will also be displayed. 
2. I see that the `self.panes_to_hide` is private in `tiled_panes`. Since I am in screen, how do I reach that? My approach is to try to create some public function in `tiled_panes` and another one in `tab`.

**Comment by imsnif** [comment]
1. Does its frame also take up half the screen? Or does the frame fill up the whole space but just the editor itself takes up half the screen?

If it's the first option, it's probably that we need to update the pty about the size change. Like we do in this macro: https://github.com/zellij-org/zellij/blob/main/zellij-server/src/tab/mod.rs#L47-L59 (though it will probably be best to do this as part of the original message exchange, since in this case we already know the size - will save a roundtrip).

If it's the second option, my guess is that the updated geom is somewhat wrong. If you log it (eg. with log::info or your favorite debugging method), does it show the half or the full geometry?

2. Yeah, that's what I would do.

**Comment by imsnif** [comment]
You can use this unwieldy method: https://github.com/zellij-org/zellij/blob/main/zellij-server/src/screen.rs#L598
That's how `Screen` gives its `os_api` to `Tab` - so should work well here.

**Comment by cosminadrianpopescu** [comment]
Trying this and I'm getting again issues with borrowing variables: 

```
error[E0502]: cannot borrow `screen.bus.os_input` as immutable because it is also borrowed as mutable
   --> zellij-server/src/screen.rs:900:48
    |
28  |                 $pane.get_content_columns() as u16,
    |                 --------------------------- mutable borrow later used here
...
891 |                 if let Some(active_tab) = screen.get_active_tab_mut(client_id) {
    |                                           ------------------------------------ mutable borrow occurs here
...
900 |                             resize_pty!(pane2, screen.bus.os_input.as_ref().unwrap().clone());
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ immutable borrow occurs here
                                                                                                                                                                               
For more information about this error, try `rustc --explain E0502`.
error: could not compile `zellij-server` due to previous error
(ins)lixa@lixa-laptop:~/programs/sources/zellij$
```

**Comment by cosminadrianpopescu** [comment]
Finally fixed and now it's fine. :) The panel has the proper size.

So, I will continue with the next steps. Thank you.

**Comment by cosminadrianpopescu** [comment]
So, now for point 7:

> For closing the pane, we'll need to keep another bit of state on Tab, maybe a HashMap<scrollback_pane_id, originating_pane_id>, so that we know to take the pane our of self.panes_to_hide instead of going through our whole normal close_pane jam.

How do I know when the panel is closed?

**Comment by imsnif** [comment]
When a pane is closed (either through a Zellij shortcut or by the program running inside the pane itself) it eventually bubbles down to here: https://github.com/zellij-org/zellij/blob/main/zellij-server/src/tab/mod.rs#L1348

**Comment by cosminadrianpopescu** [comment]
For floating panes it seems to work perfectly. 

However, there are multiple issues with the geometry when it comes to tiled panes. First approach was to save the original geometry and then to restore it in `close_pane`. This works perfectly for one pane. However, when I have multiple panes, this is what happens. 

Initially, I get this layout.

```
+------------------+------------------+
|                                     |
|                                     |
|                                     |
|                Pane 1               |
|                                     |
|                                     |
|                                     |
+------------------+------------------+
|                                     |
|                                     |
|                                     |
|                Pane 2               |
|                                     |
|                                     |
|                                     |
+------------------+------------------+
```

Then I select the second pane (Pane 2). There I do `EditScrollback`. The second pane is propely replaced with the editor. But the first panel, is also loosing the space that would've normally been occupied by the new pane (with the editor) which has the geometry of `Pane 2`. 

```
+------------------+------------------+
|                  |                  |
|                  |                  |
|                  |                  |
|      Pane 1      | Empty space      |
|                  |                  |
|                  |                  |
|                  |                  |
+------------------+------------------+
|                                     |
|                                     |
|                                     |
|         Scroll edited in editor     |
|                                     |
|                                     |
|                                     |
+------------------+------------------+
```

Im am thinking of using some strategy like when a panels is toggled as floating. But even then, probably that if I close the pane opening, then the `Pane 1` would get modified and it would take all the space. I think that we should not use the normal `new_pane` function, but instead a new one which would not process the geometry. And I see [here](https://github.com/zellij-org/zellij/blob/main/zellij-server/src/tab/mod.rs#L677) that the geometry is supposed to happen later. Where is this? 

Also, can we create a new pane without any geometry? 

Another strategy that I was thinking about was to try to somehow put it in the `floating_panes`, then give it the geometry of the panel being replaced. Like this, normally the geometry would not be modified, right? But then, what are the implications of such a thing. 

Any ideas?

**Comment by imsnif** [comment]
I agree we should not be using the new_pane method. The geometry is set in the `tiled_panes.insert_pane` method here: https://github.com/zellij-org/zellij/blob/main/zellij-server/src/panes/tiled_panes/mod.rs#L126

The `insert_pane` method basically means "Find a room for this pane and adjust the other panes accordingly". Maybe it's name is not super clear.

I think we need to use a new method in `Tab` instead of `new_pane` as you suggest that would switch the geometries, and also a new method in `tiled_panes` that would add the new pane to its state without changing the geometry. (and also we need to make sure the original pane is in `panes_to_hide` so it won't be rendered twice - but iirc you already took care of that).

**Comment by cosminadrianpopescu** [comment]
> I agree we should not be using the new_pane method. The geometry is set in the tiled_panes.insert_pane method here: https://github.com/zellij-org/zellij/blob/main/zellij-server/src/panes/tiled_panes/mod.rs#L126

I'll look into it.

> and also a new method in tiled_panes that would add the new pane to its state without changing the geometry

This new method should replace which call?

> and also we need to make sure the original pane is in panes_to_hide so it won't be rendered twice - but iirc you already took care of that

Yes, I took care of this. As I was saying, for floating panes it seems to be working without issues. I imagine that this is because when we have floating panes, inserting a new pane does not impact the geometry of others.

**Comment by imsnif** [comment]
> This new method should replace which call?

The `tiled_panes.insert_pane` call.

**Comment by cosminadrianpopescu** [comment]
So far, I have this: 

```rust
    pub fn create_pane(&mut self, pid: PaneId, client_id: Option<ClientId>, geom: PaneGeom) {
        if self.floating_panes.panes_are_visible() {
            self.new_pane(pid, client_id);
        } else {
            if self.tiled_panes.fullscreen_is_active() {
                self.tiled_panes.unset_fullscreen();
            }
            if let PaneId::Terminal(term_pid) = pid {
                let next_terminal_position = self.get_next_terminal_position();
                let new_terminal = TerminalPane::new(
                    term_pid,
                    geom, // the initial size will be set later
                    self.style,
                    next_terminal_position,
                    String::new(),
                    self.link_handler.clone(),
                    self.character_cell_size.clone(),
                    self.terminal_emulator_colors.clone(),
                    );
                self.tiled_panes.add_pane_with_existing_geom(pid, Box::new(new_terminal));
                // self.tiled_panes.insert_pane(pid, Box::new(new_terminal));
                // self.should_clear_display_before_rendering = true;
                if let Some(client_id) = client_id {
                    self.tiled_panes.focus_pane(pid, client_id);
                }
            }
        }
    }
```

This works better. There are however still issues: some crashes that I need to
investigate. Also, any idea why the frame is not drawn?

**Comment by cosminadrianpopescu** [comment]
Uff. This has other issues. If I create a pane down, I switch to the panel on the bottom, I replace with the editor, then I switch to the first panel (on top) and close it, the geometry of the panel with the buffer being edited is messed up.

**Comment by cosminadrianpopescu** [comment]
I would imagine that then one of the 2 panels (the one replacing and the one being replaced) should be excluded from any re-calculation. Any idea where I should do this? In `relayout` in `tiled_panes`?

I keep turning back to the idea of adding the new panel (the one which edits the current panel) in the list of `floating_panes`. It seems to be to be safer and pottentially create less problems. Even if let's say that I start changing the geometry of the panel editing the current buffer and then I see the buffer under it (or an empty space), is this an issue?

**Comment by imsnif** [comment]
I think we should get this to work with `TiledPanes`, exactly for this reason. The user can resize the pane while editing and we want to keep things nice and consistent. I think we can totally get it to work, it's just a little bit obtuse. Happy to take a look at the code if you want me to.

**Comment by cosminadrianpopescu** [comment]
Uploaded the code so far to https://github.com/cosminadrianpopescu/zellij. 

The branch is edit-scrollback

**Comment by imsnif** [comment]
How are you testing this? Do you maybe have a config file with the command that can save me time?

**Comment by cosminadrianpopescu** [comment]
```yaml
        - action: [EditScrollback, SwitchToMode: Locked]
          key: [Char: '?']
```

**Comment by imsnif** [comment]
Hey, I unfortunately have to go now. Things I discovered:

1. To get the pane frame to render (and generally for the pane to behave a little better) you can do (in tiled_panes/mod.rs)
```rust
    pub fn add_pane_with_existing_geom(&mut self, pane_id: PaneId, mut pane: Box<dyn Pane>) {
        if self.draw_pane_frames {
            pane.set_content_offset(Offset::frame(1));
        }
        self.panes.insert(pane_id, pane);
    }
```
I started debugging the closing issue (as well as the opening new pane while editing) and it seems to me as if tiled_panes doesn't know about this new pane somehow. It's just a feeling though, haven't had time to fully investigate it.

Will be happy to take a closer look either tonight or tomorrow morning (depending on when I get home).

**Comment by cosminadrianpopescu** [comment]
> it seems to me as if tiled_panes doesn't know about this new pane somehow. It's just a feeling though

I have the same feeling. 

Doing this:

```rust
    pub fn focus_pane(&mut self, pane_id: PaneId, client_id: ClientId) {
        self.active_panes.insert(client_id, pane_id);
        log::error!("PANES ARE: {:?}", self.active_panes);
        for (kind, pane) in self.panes.iter() {
            log::error!("KIND IS: {:?}", kind);
            log::error!("PID IS: {:?}", pane.pid());
        }
        if self.session_is_mirrored {
            // move all clients
            let connected_clients: Vec<ClientId> =
                self.connected_clients.borrow().iter().copied().collect();
            for client_id in connected_clients {
                self.active_panes.insert(client_id, pane_id);
            }
        }
    }
```

I get the following in the log:

```
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.696 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:278]: PANES ARE: {1: Terminal(16)} 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.696 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:280]: KIND IS: Terminal(9) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.697 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:281]: PID IS: Terminal(9) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.697 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:280]: KIND IS: Terminal(14) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.697 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:281]: PID IS: Terminal(14) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.697 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:280]: KIND IS: Terminal(16) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.698 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:281]: PID IS: Terminal(16) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.698 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:280]: KIND IS: Plugin(0) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.698 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:281]: PID IS: Plugin(0) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.698 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:280]: KIND IS: Plugin(1) 
ERROR  |zellij_server::panes::til| 2022-06-03 17:38:52.698 [screen    ] [zellij-server/src/panes/tiled_panes/mod.rs:281]: PID IS: Plugin(1) 
```

So, to me it looks ok. I have 3 panes in the `self.panes` (which should be fine - the top panel, bottom panel and the editor panel).

I wonder: when adding something to `panels_to_hide`, do I need to remove it from `panes`?

**Comment by imsnif** [comment]
Hey! So got a chance to take a deeper look at this.

Your instincts were right: this (and some other problems) are happening because `TiledPanes` mostly ignores panes_to_hide. The only occasion we've been using it up until now is for placing a pane in fullscreen - which skips alot of the logic we do need here.

I have a partial fix for this locally and will push it to your branch once its done (I need to leave the house for a couple of hours, will wrap it up afterwards).

In the interests of wrapping this up in time, would you like to meanwhile spend time wrapping up other parts of this PR? eg. writing tests? Or anything else you think you needs adding (I think you have more context here than me)?

**Comment by imsnif** [comment]
Also - one more thing I see we need to do is make sure (as much as possible) that the editor will be opened to the same line in the scrollbuffer

**Comment by imsnif** [comment]
Also there's a funny feature that I'm not sure if we should disable or not: you are totally able nest these. If I open the inplace editor while it is already opened in the same pane, I see the textual representation of my vim session. :D

**Comment by cosminadrianpopescu** [comment]
> In the interests of wrapping this up in time, would you like to meanwhile spend time wrapping up other parts of this PR? eg. writing tests? Or anything else you think you needs adding (I think you have more context here than me)?

Yes, I will try to write a test case. In my vision, a proper test case should also test splitting the panes, should test the floating panes and should also test different layouts. The problem is that I probably don't have time to implement this this evening (because tomorrow I don't think I will have time to put into it). 

What I would probably have time to implement this evening, is a basic test case: edit the buffer, assert that the editor is opened and assert that the editor content is the buffer content (assuming that this is straight forward in testing the `OpenFile` command). That to me would be a basic test case. 

So, would it be ok with a basic test case now and a proper test case in the future?

And also, to me, having seen that there are quite some issues with putting a pane in `panes_to_hide`, I would've liked to have at least one week of testing this in prod. My plan was to develop this until Friday evening and then test it next week at work. Of course, this I could not manage, unfortunatelly. I suppose it depends on how sure you are on the fixes that you are doing.

> Or anything else you think you needs adding (I think you have more context here than me)?

Yes, I would like to also add to this feature 2 things: the possibility to set the `${EDITOR}` command parameters (which would answer also your other question of going to the last line). Of course, we can't just go to the last line, because the default editor can be `emacs`, or `nano`, or even who knows what graphical editor. So this should be done via an option in the config file. 

Another option I would see as a plus to this feature, would be to be able to choose the folder where we want to save the file. You saw that I'm generating `.dump` files and by default will fill the temp folder with a lot of those files. It would be nice if those files could be greoupped somewhere. 

But again, I will probably not have time to implement this features this evening. 

So, my proposal is this: if you want to release next week (again, depending on how sure you are of the fix that you are adding), let's go with a basic test case and without the other 2 features. Being able to edit the scroll back in your editor of choice is anyway a killer feature even without going to the last line and filling up the `tmp` folder with `.dump` files. We can add those afterwards, together with a proper test case.

**Comment by imsnif** [comment]
Hey @cosminadrianpopescu - so I just pushed the fixes to the branch.

There is another small bug that I was hoping you could take a look at:
1. Open an editor window
2. resize it
3. close it
The previous pane comes back with the wrong geometry (before the resize).

To fix this, I think we need (when closing the editor pane) to take its existing geometry and set it on the suppressed pane that we're now showing.

How about you take care of this and I'll do all the rest? I know the tests well so can whip up some good test cases in a short time.

About the lines I really want this to be a best effort sort of thing by default. Like "if I'm vim, use this flag, if I'm emacs use this flag, if I'm nano, use this flag - otherwise don't do it". Then in the future if you want to add the command to the config, that'll be killer. I'm also happy to take care of this.

What do you think?

**Comment by cosminadrianpopescu** [comment]
> How about you take care of this and I'll do all the rest?

Sounds perfect. I hope I can do it this evening.

**Comment by cosminadrianpopescu** [comment]
I've pushed the changes to fix the bug that you've found, but I found some issues with the floating panes:

When you edit a floating pane, then (with the edited buffer editor still open), you open another buffer. Then when you cycle through the buffers, you can reach the hidden one. Since the buffer has been added to hidden panels, I think also the floating panels should take the hidden panels into consideration (basically same fix you did for tiled panes. But the issue is that the tiled panes don't have a panes_to_hide (btw, with the last fix, I've also stop adding the panel to the hidden panels if the floating panels are visible. 

This I'm not sure if it's an issue or not (because basically the new panel just takes up the same geometry as the panel being edited). 

But the second issue is more complicated:

Open the floating panels, edit the scrollback buffer, then embed the panel. After closing the editor, the edited buffer is not retrieved, because it has not been embeded. Of course, you can retrieve it by showing again the floating panels. Is this an issue?

**Comment by imsnif** [comment]
Hum... I see what you mean. Good finds!
I think the best way to fix these (especially the second issue) is a new approach: instead of putting the panes in `panes_to_hide`, I think we should remove the `Pane` object entirely from either `TiledPanes` or `FloatingPanes`, place it in `Tab` in a new `suppressed_panes: HashMap<PaneId, Box<dyn Pane>>`. Then when the editor pane is closed, we reach for that Pane, change its geom to the one of the closed pane and place it back in either `FloatingPanes` if they are showing or `TiledPanes` if not.
If you don't have the time for this, I can add it to my list. As you prefer.

**Comment by cosminadrianpopescu** [comment]
I've thought at these issues. I think I could do the following:

When you edit a floating pane, I can move it from floating panes to tiles panes in the list of hidden panes. Then, when you close the editor, if the editor was embedded, then the buffer will be displayed properly (so nothing special todo). If the buffer was still floating, then move the rlaced buffer back from tiled to floating panes and focus it.

What so you think?

**Comment by imsnif** [comment]
Sure, that can work. Personally I prefer what I suggested above with the suppressed_panes - I think it achieves the same and is a little simpler to understand - but as you wish.

**Comment by cosminadrianpopescu** [comment]
> Sure, that can work. Personally I prefer what I suggested above with the suppressed_panes - I think it achieves the same and is a little simpler to understand 

Sure, it seems clear. But if I implement this, it should also work for tiled panes, right? I mean, we don't need to put it in hidden panes, right? Or we implement this only for floating panes?

**Comment by cosminadrianpopescu** [comment]
> it should also work for tiled panes, right? I mean, we don't need to put it in hidden panes, right? Or we implement this only for floating panes?

Forget what I said. :D I see that you already suggested this. :)

**Comment by cosminadrianpopescu** [comment]
I've ran into another of rust quirks. For sure, this language is not for me. Fighting with it to just do some basic stuff, like sending a reference from here to there... So, here we go. I do this:

```rust
                    if let Some(replaced_pane) = self.suppressed_panes.get(&(*info).pid) {
                        self.tiled_panes.add_pane_with_existing_geom(replaced_pane.pid(), replaced_pane);
                        self.suppressed_panes.remove(&(*info).pid);
                    }
```

which give me this error: 

```
1407 |                         self.tiled_panes.add_pane_with_existing_geom(replaced_pane.pid(), replaced_pane);
     |                                                                                           ^^^^^^^^^^^^^ expected struct `Box`, found reference

```

which I understand. So, for me the resolve would be to do `self.tiled_panes.add_pane_with_existing_geom(replaced_pane.pid(), *replaced_pane);`, which I did in other parts (I put the `*` to access the value from the given reference). 

But now I get this error:

```
1407 |                         self.tiled_panes.add_pane_with_existing_geom(replaced_pane.pid(), *replaced_pane);
     |                                                                                           ^^^^^^^^^^^^^^ move occurs because `*replaced_pane` has type `Box<dyn tab::Pane>`, which does not implement the `Copy` trait
```

which I kinda understand, but which I don't know how to solve.

**Comment by cosminadrianpopescu** [comment]
Found the solution, for the previous problem.

**Comment by cosminadrianpopescu** [comment]
Check out the last commit. It implements the new strategy that you suggested, but it has a big issue:

Normally, on the first glance it works perfectly, including also the floating_panes. But...

If I do for example `tail -f /tmp/ffile`, then I edit the buffer, and then, from another pane I change the content of the `/tmp/file` file, the pane does not get updated. 

Do you think my strategy (of keeping the pane always in tiled_panes and hidden_panes would prevent this?

**Comment by cosminadrianpopescu** [comment]
So, what I would expect would be this:

```
tail -f /tmp/file
```

The output of the pane is this:

```
file content line 1
```

Then, I launch `EditScrollback`. A vim editor opens with `file content line 1`. 

Then, I add another pane and do `echo "file content line 2" >> /tmp/file`

Then I go back to the pane where the buffer is edited and I close the editor. I would expect now to see on the screen

```
file content line 1
file content line 2
```

But I only see the first line. More, then if I go again to a new pane and do `echo "file content line 3" >> /tmp/file, I would expect to see on the screen 

```
file content line 1
file content line 2
file content line 3
```

But what I actually see is 

```
file content line 1
file content line 3
```

that is actually wrong. 

So, do you think that my strategy could solve this issue?

**Comment by cosminadrianpopescu** [comment]
I had another 2 commits, one reverting the previous one and another one implementing the strategy I mentioned. It works perfectly now for tiled_panes (meaning it's not losing any content) but for floating panes only issue 1 is solved (you can't scroll through the pane). The second issue is not yet solved, althought I don't understand why. If you open a floating pane, edit the buffer and then you embed it and then you close it, it will not be replaced by the old panel. The old panel you can retrieve it by showing again the floating panes. What am I missing? Because normally, the pane being already in the tiled_panes and getting the geometry of the editor pane, it should be displayed in it's place, right?

**Comment by imsnif** [comment]
Hey @cosminadrianpopescu - sorry for not responding, fell asleep quite early.
I'll take a look at all of those issues today and bring this to the finishing line.

I know you're busy today, but if I could grab just a few moments on your time to help me out, could you:
1. Open a draft PR to the main repo so that you'll get credit for these changes? (you did most of the work after all!)
2. I know the reason my strategy with the `suppressed_panes` was losing content and it's an easy fix. Other than that, are there other issues there?

**Comment by cosminadrianpopescu** [comment]
Found the problem. Although the pane is embedded, the `self.floating_panes.move_clients_out_of_pane(id)` still returns true. So, is this a bug? Because if not, then the behaviour is completely normal. It means that this is how `zellij` is supposed to behave.

**Comment by cosminadrianpopescu** [comment]
> I know the reason my strategy with the suppressed_panes was losing content and it's an easy fix. Other than that, are there other issues there?

So, what do you advise? I should go back to commit `c52a203a20cf4c87c147be8b9c193ed6458c1038` to put back your strategy? Because right now, I put in place the other strategy. 

I will open the PR.

**Comment by cosminadrianpopescu** [comment]
But before opening the PR, let me know how do you want me to open? Where it is now or at the commit `c52a203a20cf4c87c147be8b9c193ed6458c1038`?

**Comment by imsnif** [comment]
Open it where it is now. I'll take a look at this strategy and only change it if there's a need. I'm a little worried that it's a bit complex (I know I'll forget why we did this in 2 months :) ) - but let's see.

**Comment by imsnif** [comment]
Just pushed my fixes from yesterday. These:
1. Open editors to the right line based on the filename (turns out the argument is identical to all major editors, which is pretty cool). Anyway, if they don't succeed with the argument, the try to open the editor without.
2. Handles errors for a non-existent `$EDITOR` and `$VISUAL`. The user won't get a visual indication (because we sadly don't have an easy way of doing this yet), but the error - along with how to fix it - will appear in the logs.

Going to take a look at your changes now, see what needs to be fixed, then write tests, do some cleanup and we're good I think (unless I'm forgetting something).

**Comment by cosminadrianpopescu** [comment]
PR opened. 

The only minor issues I see are those I've told you about plus another one, also with floating panes. If you open a floating pane, you then edit the scroll buffer, then you embed and then you make float again, you will see both of the panes. This is again, because of `self.floating_panes.panes_contain(&id)` even when the pane is embedded. And if this is not a bug, then I would say that the behavior is normal, in my oppinion. 

What do you think?

**Comment by imsnif** [comment]
Hum... does this happen with a regular non-embedded editor pane? I don't fully have context on this yet. Just diving in now to your changes.

Can we sum up the issues that we have with the current implementation?

**Comment by cosminadrianpopescu** [comment]
So, far there are 2 minor issues, in my oppinion: 

1. Open floating panes. In the opened floating panel, edit the scroll back. Now your buffer is replaced by the editor. Now embed the panel, now close the editor. You won't see the panel non-embeded. Now, in my oppinion, if `self.floating_panes.panes_contain(&id)` should return `false` when the panel `id` is embedded, then this is a bug. But the bug has to be corrected by correcting the `self.floating_panes.panes_contain` to return false when the pane `id` is embedded. Then also this feature should work properly. If `self.floating_panes.panes_contain(&id)` should return true even if the panel is embedded, then it's normal that after I embeded the editor and closed it I should again open the floating panes to see the edited buffer. Maybe do this automatically in the close_pane function?
2. Same scenario: Open floating panes, in the opened floating panel edit the scroll back. Now embed the panel and now make the editor panel floating again. At this point, you will see the 2 panes (the editor and the edited panel). Here, I am not that sure that even if `self.floating_panes.panes_contain(&id)` would return `false` it would work. Or even that it should not work if it returns `true`. 

Unfortunatelly I don't have discord or any other social app (WhatsApp, Teams etc.). I only have signal messenger. If you want, you can send me a private message at `cosminadrianpopescu` `at` `gmail.com` and we can discuss directly via `Signal Messenger`, if you have it.

**Comment by cosminadrianpopescu** [comment]
If you don't have signal messenger, I have my own infrastructure at `https://cloud.taid.be` (my company) where we can do video calls and all the others. Again, give me a message on my private e-mail and I will send you a link, where you can connect and we can talk directly (video and audio or only audio) and we can share screens and everything that you would expect from such a thing.

**Comment by imsnif** [comment]
Hey @cosminadrianpopescu - I pushed changes into the branch that should fix everything.

I hope you don't mind, but I ended up using the `suppressed_panes` strategy. I just know this part of the code base very well and know what pitfalls to avoid. There would just be too many things to fix otherwise (as you started seeing with the bugs you mentioned).

This fixed all the bugs you mentioned - including the one with new info coming to the suppressed terminal (eg. with tail). But please check me to make sure I'm not mistaken (when you have the time).

I'm taking a lunch break and afterwards I'm going to write tests and maybe look into placing an editor command override in the config as you suggested (if it won't take too long).

I can wait until tomorrow to merge this if you don't have time to test today.

**Comment by cosminadrianpopescu** [comment]
> I hope you don't mind, but I ended up using the suppressed_panes strategy. 

No, of course. Whatever works best and faster.

> But please check me to make sure I'm not mistaken (when you have the time).

I will check out a little bit later. Right now I'm not at a computer.

**Comment by cosminadrianpopescu** [comment]
All works perfect now (floating and tiled panes). I did not find any new issues. I will keep testing and starting Tuesday, I will use it also in a prod like environment (I will build the new version at work) and let you know if I found more issues (I hope not).  

I was thinking of one last nice touch. To name the panel rather than "Panel #<n>", maybe we can give it a nice name, like "Edit Panel #<n>" which would then be the number of the panel edited? Or if the panel edited has a given name, then it would be "Edit Panel Custom Name". 

How difficult would that be? I might be able to give it one hour this evening later. Do you think is fesable? Or better to leave it for later?

**Comment by imsnif** [comment]
Nice! I plan on releasing either tomorrow or on Tuesday, so you can already use the release version and if there's a bug we can always release a patch.

I like the title idea. I'd call it "Editing Pane # Scrollback" 
I think it shouldn't be too difficult if we pass the title explicitly here: https://github.com/cosminadrianpopescu/zellij/blob/edit-scrollback/zellij-server/src/tab/mod.rs#L714
The problem is that this will always override whatever title the editor sets... Not too bad though - we can work around it later and I think this is a cool default.

**Comment by imsnif** [comment]
You might need to update some snapshots since I already added tests, but shouldn't be too difficult with `cargo insta review`.

**Comment by cosminadrianpopescu** [comment]
I will give it a try and let you know if I can come up with anything. If not, I suppose I can add it at a later stage after release, right? Since today I'm not that free.

**Comment by imsnif** [comment]
For sure, no stress! As you said, it's a pretty amazing feature as is. I'm very much looking forward to using this in my day-to-day.

**Comment by imsnif** [comment]
Alright, all is done in the PR.
I added:
1. Tests (several integration to test edge cases and one e2e to test everything together)
2. Some minor fixes here and there
3. A config attribute: `scrollback_editor` to allow overriding the environment variable
4. A section in `zellij setup --check` that let's the user know which editor command we're going to use

I also merged from main. So @cosminadrianpopescu - if you want to add the window title sometime tonight, please do go ahead. Otherwise I'll merge this to main tomorrow and release either tomorrow or on Tuesday.

**Comment by cosminadrianpopescu** [comment]
No, unfortunately I didn't had any time this evening. I'll add that with other release.

**Comment by nlvw** [comment]
If the scrollback_editor option or `$EDITOR` env variable has any flags then it fails.  This is the case wether there is an empty space at the end or not in the config file.

```console
$ echo $EDITOR
vim -u ~/.config/vim/vimrc.vim
```

```yaml
scrollback_editor: '/usr/bin/less --use-color '
```

**Comment by cosminadrianpopescu** [comment]
> Does it work if you wrap it all in bash script to get the flags?

I've had the same issue as @nlvw and I can confirm that it works if you wrap it in a script and point the `$EDITOR` variable to that script.

**Comment by imsnif** [comment]
Question is... is this a bug or the intended behaviour? I wonder if eg. git picks up the $EDITOR flags when editing merge commits and rebases...

**Comment by cosminadrianpopescu** [comment]
I found a problem with the current implementation of the feature:

If you have several split panes, then you full screen one, then you replace the editor scrollback and then you close the editor, the pane will get the geometry of the buffer at exit point, which due to full screen was the max geometry. So, when you take the pane out of the full screen, then it will remain maximized. I will look into it.

**Comment by imsnif** [comment]
I'm closing this issue for now though because the feature itself was already implemented. @cosminadrianpopescu - please keep me updated on the issue. We can always open a new one.
